cpp/src/explainer/tree_shap.cu

```diff
@@ -159,50 +114,90 @@ class DenseDatasetWrapper {
 };
 
+template <typename ThresholdType>
+void gpu_treeshap_impl(const TreePathInfoImpl<ThresholdType>* path_info,
+                       const float* data,
+                       std::size_t n_rows,
+                       std::size_t n_cols,
+                       float* out_preds)
+{
+  DenseDatasetWrapper X(data, n_rows, n_cols);
+
+  std::size_t num_groups = 1;
+  if (path_info->task_type == tl::TaskType::kMultiClfGrovePerClass &&
+      path_info->task_param.num_class > 1) {
+    num_groups = static_cast<std::size_t>(path_info->task_param.num_class);
+  }
+  std::size_t pred_size = n_rows * num_groups * (n_cols + 1);
+
+  thrust::device_ptr<float> out_preds_ptr = thrust::device_pointer_cast(out_preds);
+  gpu_treeshap::GPUTreeShap(X,
+                            path_info->paths.begin(),
+                            path_info->paths.end(),
+                            num_groups,
+                            out_preds_ptr,
+                            out_preds_ptr + pred_size);
+
+  // Post-processing
+  auto count_iter  = thrust::make_counting_iterator(0);
+  auto num_tree    = path_info->num_tree;
+  auto global_bias = path_info->global_bias;
+  if (path_info->average_tree_output) {
+    thrust::for_each(
+      thrust::device, count_iter, count_iter + pred_size, [=] __device__(std::size_t idx) {
+        out_preds[idx] /= num_tree;
+      });
+  }
+  thrust::for_each(
+    thrust::device,
+    count_iter,
+    count_iter + (n_rows * num_groups),
+    [=] __device__(std::size_t idx) { out_preds[(idx + 1) * (n_cols + 1) - 1] += global_bias; });
+}
 
 }  // anonymous namespace
 
 namespace ML {
 namespace Explainer {
 
 template <typename ThresholdType, typename LeafType>
-ExtractedPathHandle
-extract_paths_impl(const tl::ModelImpl<ThresholdType, LeafType>& model) {
+std::unique_ptr<TreePathInfo> extract_path_info_impl(
+  const tl::ModelImpl<ThresholdType, LeafType>& model)
+{
   if (!std::is_same<ThresholdType, LeafType>::value) {
     RAFT_FAIL("ThresholdType and LeafType must be identical");
   }
-  if (model.task_type != tl::TaskType::kBinaryClfRegr
-      && model.task_type != tl::TaskType::kMultiClfGrovePerClass) {
-    RAFT_FAIL("Only tree models with task type kBinaryClfRegr, "
-              "kMultiClfGrovePerClass are supported. So for example, "
-              "scikit-learn trees are not supported.");
+  if (model.task_type != tl::TaskType::kBinaryClfRegr &&
+      model.task_type != tl::TaskType::kMultiClfGrovePerClass) {
+    RAFT_FAIL("cuML RF / scikit-learn classifiers must have n_classes == 2");
   }
-  std::unique_ptr<ExtractedPath> path_container
-    = ExtractedPath::Create<ThresholdType>();
-  ExtractedPathImpl<ThresholdType>* paths
-    = dynamic_cast<ExtractedPathImpl<ThresholdType>*>(path_container.get());
+  std::unique_ptr<TreePathInfo> path_info_ptr = std::make_unique<TreePathInfoImpl<ThresholdType>>();
+  auto* path_info = dynamic_cast<TreePathInfoImpl<ThresholdType>*>(path_info_ptr.get());
```

Comment:
I think it would be better to avoid the `dynamic_cast` here (and the general complexity of the inner implementation classes for different types) by using a single struct with a `std::variant` member variable, but let's just make note of that and save it for a follow-on.